straxen/plugins/event_patternfit.py

```diff
@@ -0,0 +1,410 @@
+    if (k < 0):
+        return np.nan
+    if (k == n):
+        return (1.0)
+    dn = n - k
+    if (k == 0):
+        dk = np.exp(dn * np.log(1.0 - p))
+    else:
+        dk = k + 1
+        dk = betainc(dn, dk, 1.0 - p)
+    return dk
+
+# Continuos binomial distribution
+def binom_pmf(k, n, p):
+    scale_log = gammaln(n + 1) - gammaln(n - k + 1) - gammaln(k + 1)
+    ret_log = scale_log + k * np.log(p) + (n - k) * np.log(1 - p)
+    return np.exp(ret_log)
+
+def binom_cdf(k, n, p):
+    return bdtr(k, n, p)
+
+def binom_sf(k, n, p):
+    return bdtrc(k, n, p)
+
+def binom_test(k, n, p):
+    '''
+    The main purpose of this algorithm is to find the value j on the
+    other side of the mean that has the same probability as k, and
+    integrate the tails outward from k and j. In the case where either
+    k or j are zero, only the non-zero tail is integrated.
+    '''
+    d = binom_pmf(k, n, p)
+    rerr = 1 + 1e-7
+    d = d * rerr
+    # define number of intereation for finding the the value j
+    # the exeptional case of n<=0, is avoid since n_iter is at least 2
+    n_iter = int(max(np.round(np.log10(n)) + 1, 2))
+
+    if k < n * p:
+        # if binom_pmf(n, n, p) > d, with d<<1e-3, means that we have
+        # to look for j value above n. It is likely that the binomial
+        # test for such j is extremely low such that we can stop
+        # the algorithm and return 0
+        if binom_pmf(n, n, p) > d:
+            for n_ in np.arange(n, 2*n, 1):
+                if binom_pmf(n_, n, p) < d:
+                    j_min, j_max = k, n_
+                    do_test = True
+                    break
+                do_test = False
+        else:
+            j_min, j_max = k, n
+            do_test = True
+        def _check_(d, y0, y1):
+            return (d>y1) and (d<=y0)
+    else:
+        if binom_pmf(0, n, p) > d:
+            n_iter, j_min, j_max = 0, 0, 0
+        else:
+            j_min, j_max = 0, k
+        do_test = True
+        def _check_(d, y0, y1):
+            return (d>=y0) and (d<y1)
+
+    # if B(k;n,p) is already 0 or I can't find the j in the other side of the mean
+    # the returned binomial test is 0
+    if (d==0)|(not do_test):
+        pval = 0
+    else:
+        # Here we are actually looking for j
+        for i in range(n_iter):
+            n_pts = int(j_max - j_min)
+            j_range = np.linspace(j_min, j_max, n_pts, endpoint=True)
+            y = binom_pmf(j_range, n, p)
+            for i in range(len(j_range) - 1):
+                if _check_(d, y[i], y[i + 1]):
+                    j_min, j_max = j_range[i], j_range[i + 1]
+                    break
+        j = max(min((j_min + j_max) / 2, n), 0)
+
+        # One side or two side
+        # binomial test
+        if k * j == 0:
+            pval = binom_sf(max(k, j), n, p)
+        else:
+            pval = binom_cdf(min(k, j), n, p) + binom_sf(max(k, j), n, p)
+        pval = min(1.0, pval)
+        
+    return pval
```

Comment:
Can't we use scipy here?

Reply:
You mean the [scipy.stats.binom_test](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.binom_test.html) ? No we can't. The package in scipy is defined such that is valid for integer number. Here we need the continuos version, like in [XENON1T](https://github.com/XENON1T/pax/blob/master/pax/plugins/interaction_processing/S1AreaFractionTopProbability.py)

Reply:
Thanks! Perhaps it's worth considering adding to strax.processing as it's so general.